crates/rpc/rpc-builder/src/rate_limiter.rs

```diff
@@ -0,0 +1,111 @@
+use jsonrpsee::{server::middleware::rpc::RpcServiceT, types::Request, MethodResponse};
+use reth_tasks::pool::BlockingTaskGuard;
+use tokio::sync::OwnedSemaphorePermit;
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower::Layer;
+
+/// Rate limiter for the RPC server.
+///
+/// Rate limits expensive calls such as debug_ and trace_.
+#[derive(Debug, Clone)]
+pub(crate) struct RpcRequestRateLimiter {
+    inner: Arc<RpcRequestRateLimiterInner>,
+}
+
+impl RpcRequestRateLimiter {
+    pub(crate) fn new(handle: tokio::runtime::Handle, rate_limit: usize) -> Self {
+        Self {
+            inner: Arc::new(RpcRequestRateLimiterInner {
+                handle,
+                call_guard: BlockingTaskGuard::new(rate_limit),
+            }),
+        }
+    }
+}
+
+impl<S> Layer<S> for RpcRequestRateLimiter {
+    type Service = RpcRequestRateLimitingService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        RpcRequestRateLimitingService::new(inner, self.clone())
+    }
+}
+
+/// Rate Limiter for the RPC server
+#[derive(Debug)]
+struct RpcRequestRateLimiterInner {
+    call_guard: BlockingTaskGuard,
+    handle: tokio::runtime::Handle,
+}
```

Comment:
to use `Arc<Semaphor>` in poll API we can either use a `Box<dyn Future>` as the Future type, but there's also 

https://docs.rs/tokio-util/latest/tokio_util/sync/struct.PollSemaphore.html

for this use case